File: packages/depz-sensor-sdk-csharp/src/Depz.Sensor/Dataset/Dataset.cs
using System.IO.Compression;
using System.Text;
using System.Text.Json;

namespace Depz.Sensor.Dataset;

/// <summary>Per-device time-sync (contract 02 §5): device_clock − host_clock, and RTT.</summary>
public sealed record TimeSync(long OffsetUs, long RttUs);

/// <summary>Header metadata for one device in a dataset (contract 09).</summary>
public sealed record DatasetDevice(
    string Id, string? Serial, string? SoftwareName, string? SensorType, TimeSync? TimeSync);

/// <summary>
/// One decoded record on the shared host timeline. <see cref="Value"/> is the
/// kind-specific payload object (players use the fields present for the kind).
/// </summary>
public sealed record DatasetRecord(string DeviceId, long THostUs, string Kind, JsonElement Value);

/// <summary>
/// Reader for the <c>.depzdata</c> decoded, multi-device, time-synced dataset
/// format (contracts/09_DATASET_FORMAT.md). Plain or gzip (<c>.depzdata.gz</c>).
/// Records are exposed merged by host time <c>t</c>; unknown kinds are kept.
/// </summary>
public sealed class DatasetReader
{
    /// <summary>Device metadata from the header, keyed by device id (d0, d1, …).</summary>
    public IReadOnlyDictionary<string, DatasetDevice> Devices { get; }

    /// <summary>All records, stably merged by ascending host time.</summary>
    public IReadOnlyList<DatasetRecord> Records { get; }

    public string? Note { get; }

    public DatasetReader(string path)
        : this(ReadAllLines(path))
    {
    }

    private DatasetReader(IReadOnlyList<string> lines)
    {
        if (lines.Count == 0)
            throw new InvalidDataException("empty dataset (no header line)");

        using JsonDocument headerDoc = JsonDocument.Parse(lines[0]);
        JsonElement header = headerDoc.RootElement;

        var devices = new Dictionary<string, DatasetDevice>();
        if (header.TryGetProperty("devices", out JsonElement devs) && devs.ValueKind == JsonValueKind.Object)
        {
            foreach (JsonProperty dev in devs.EnumerateObject())
            {
                JsonElement m = dev.Value;
                TimeSync? sync = null;
                if (m.TryGetProperty("time_sync", out JsonElement ts) && ts.ValueKind == JsonValueKind.Object)
                {
                    sync = new TimeSync(
                        ts.GetProperty("offset_us").GetInt64(),
                        ts.TryGetProperty("rtt_us", out JsonElement rtt) ? rtt.GetInt64() : 0);
                }
                devices[dev.Name] = new DatasetDevice(
                    dev.Name,
                    GetStringOrNull(m, "serial"),
                    GetStringOrNull(m, "software_name"),
                    GetStringOrNull(m, "sensor_type"),
                    sync);
            }
        }
        Devices = devices;
        Note = GetStringOrNull(header, "note");

        var records = new List<(long T, int Order, DatasetRecord Rec)>();
        int order = 0;
        for (int i = 1; i < lines.Count; i++)
        {
            string line = lines[i];
            if (line.Length == 0)
                continue;
            using JsonDocument doc = JsonDocument.Parse(line);
            JsonElement r = doc.RootElement;
            long t = r.GetProperty("t").GetInt64();
            var rec = new DatasetRecord(
                r.GetProperty("d").GetString()!,
                t,
                r.GetProperty("k").GetString()!,
                r.TryGetProperty("v", out JsonElement v) ? v.Clone() : default);
            records.Add((t, order++, rec));
        }
        // Merge by host time; stable within equal t (append order preserved).
        records.Sort((a, b) =>
        {
            int c = a.T.CompareTo(b.T);
            return c != 0 ? c : a.Order.CompareTo(b.Order);
        });
        Records = records.Select(x => x.Rec).ToList();
    }

    private static string? GetStringOrNull(JsonElement obj, string name) =>
        obj.TryGetProperty(name, out JsonElement e) && e.ValueKind == JsonValueKind.String
            ? e.GetString()
            : null;

    private static IReadOnlyList<string> ReadAllLines(string path)
    {
        byte[] raw = File.ReadAllBytes(path);
        // gzip magic (also covers .depzdata.gz files without extension check).
        if (raw.Length >= 2 && raw[0] == 0x1F && raw[1] == 0x8B)
        {
            using var ms = new MemoryStream(raw);
            using var gz = new GZipStream(ms, CompressionMode.Decompress);
            using var reader = new StreamReader(gz, Encoding.UTF8);
            raw = Encoding.UTF8.GetBytes(reader.ReadToEnd());
        }
        string text = Encoding.UTF8.GetString(raw);
        return text.Split('\n').Where(l => l.Trim().Length > 0).ToArray();
    }
}

File: packages/depz-sensor-sdk-java/src/main/java/ai/depz/sensor/dataset/Dataset.java
package ai.depz.sensor.dataset;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * {@code .depzdata} dataset reader (contracts/09_DATASET_FORMAT.md): a JSONL
 * file whose first line is the header ({@code schema}, {@code devices}, ...)
 * and whose remaining lines are records {@code {"d","t","k","v"}}. Records are
 * stably merged onto one host timeline by {@code t} (µs). Mirror of the
 * TS/Python {@code DatasetReader}.
 *
 * <p>JSON parsing is delegated to a caller-supplied line parser so this module
 * has no dependency on the test harness's JSON code; see {@link #parse}.
 */
public final class Dataset {
    public static final String DATASET_SCHEMA_PREFIX = "depz.dataset/";

    /** One merged record. */
    public record Record(String deviceId, long tHostUs, String kind, Map<String, Object> value) {}

    private final Map<String, Object> header;
    private final List<Record> records;

    private Dataset(Map<String, Object> header, List<Record> records) {
        this.header = header;
        this.records = records;
    }

    /**
     * Parse dataset content. {@code jsonLine} converts one JSON line to a
     * {@code Map<String,Object>} (objects → Map, arrays → List, ints → Long).
     */
    public static Dataset parse(String content, java.util.function.Function<String, Object> jsonLine) {
        List<String> lines = new ArrayList<>();
        for (String l : content.split("\n")) {
            if (!l.trim().isEmpty()) {
                lines.add(l);
            }
        }
        if (lines.isEmpty()) {
            throw new IllegalArgumentException("empty dataset");
        }
        @SuppressWarnings("unchecked")
        Map<String, Object> header = (Map<String, Object>) jsonLine.apply(lines.get(0));
        Object schema = header.get("schema");
        if (!(schema instanceof String s) || !s.startsWith(DATASET_SCHEMA_PREFIX)) {
            throw new IllegalArgumentException("not a depz.dataset file");
        }
        // Preserve file order for the stable sort tiebreak.
        List<Record> raw = new ArrayList<>();
        for (int i = 1; i < lines.size(); i++) {
            @SuppressWarnings("unchecked")
            Map<String, Object> ev = (Map<String, Object>) jsonLine.apply(lines.get(i));
            @SuppressWarnings("unchecked")
            Map<String, Object> v = (Map<String, Object>) ev.get("v");
            raw.add(new Record(
                    (String) ev.get("d"), ((Number) ev.get("t")).longValue(), (String) ev.get("k"), v));
        }
        // Per-device order is monotonic; a stable sort by tHostUs merges exactly.
        List<Integer> idx = new ArrayList<>();
        for (int i = 0; i < raw.size(); i++) {
            idx.add(i);
        }
        idx.sort((a, b) -> {
            int c = Long.compare(raw.get(a).tHostUs(), raw.get(b).tHostUs());
            return c != 0 ? c : Integer.compare(a, b);
        });
        List<Record> sorted = new ArrayList<>();
        for (int i : idx) {
            sorted.add(raw.get(i));
        }
        return new Dataset(header, sorted);
    }

    public Map<String, Object> header() {
        return header;
    }

    public String schema() {
        return (String) header.get("schema");
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> devices() {
        Object d = header.get("devices");
        return d == null ? Map.of() : (Map<String, Object>) d;
    }

    public List<Record> records() {
        return records;
    }

    public long durationUs() {
        if (records.isEmpty()) {
            return 0;
        }
        return records.get(records.size() - 1).tHostUs() - records.get(0).tHostUs();
    }
}
